system/jlib/jstring.cpp

```diff
@@ -1063,13 +1063,48 @@ StringBuffer & StringBuffer::replaceString(const char* oldStr, const char* newSt
     if (curLen && oldStr)
     {
         size_t oldlen = strlen(oldStr);
-        if (oldlen > curLen)
-            return *this;
-
-        StringBuffer temp;
         size_t newlen = newStr ? strlen(newStr) : 0;
-        if (::replaceString(temp, curLen, buffer, oldlen, oldStr, newlen, newStr, true))
-            swapWith(temp);
+        // Search string (oldStr) cannot be larger than the source string
+        if (oldlen <= curLen)
+        {
+            // If target string (newStr) is shorter than or equal to search string, do an in-place replacement to avoid allocation
+            if (oldlen >= newlen)
+            {
+                size_t maxOffset = curLen - oldlen;
+                size_t offset = 0;
+                size_t targetOffset = 0;
+                size_t lastCopied = 0;
+                char firstChar = oldStr[0];
+                while (offset <= maxOffset)
+                {
+                    if (unlikely(buffer[offset] == firstChar)
+                        && unlikely((oldlen == 1) || memcmp(buffer + offset, oldStr, oldlen)==0))
+                    {
+                        if (lastCopied)
```

Comment:
`if (lastCopied != targetOffset)
`
Protects against copying to the same location when newlen=oldlen
`  && likely(lastCopied != offset)
`
Protect against zero length copy.